eth/backend.go

```diff
@@ -776,19 +781,29 @@ func (s *Ethereum) StartMining(ctx context.Context, db kv.RwDB, mining *stagedsy
 		defer debug.LogPanic()
 		defer close(s.waitForMiningStop)
 
-		mineEvery := time.NewTicker(3 * time.Second)
+		mineEvery := time.NewTicker(cfg.Recommit)
 		defer mineEvery.Stop()
 
+		// Listen on a new head subscription. This allows us to maintain the block time by
+		// triggering mining after the block is passed through all stages.
+		newHeadCh, closeNewHeadCh := s.notifications.Events.AddHeaderSubscription()
+		defer closeNewHeadCh()
+
 		var works bool
 		var hasWork bool
 		errc := make(chan error, 1)
 
 		for {
-			mineEvery.Reset(3 * time.Second)
+			mineEvery.Reset(cfg.Recommit)
 			select {
 			case <-s.notifyMiningAboutNewTxs:
+				log.Debug("Start mining new block based on txpool notif")
+				hasWork = true
+			case <-newHeadCh:
+				log.Debug("Start mining new block based on new head channel")
```

Comment:
do you wan't only start mine here? Or also stop current mining progress?

Reply:
I think this case will occur when a block has been written to db, which means that it's past the mining stage and the mining process is already stopped. I am not completely sure. What do you suggest? Also, how do you signal mining to stop?

Reply:
2 use-cases:
- you are only miner in network 
- there are many other miners in network. In this case you may receive blocks mined not by you, while your mining is still in-progress.

Reply:
Okay makes sense. How to tackle this?

Reply:
Also, is it the case that the moment when we're receiving blocks, it would stop mining? 
I know that geth listens to the downloader event as well to toggle mining. What do you think?

Reply:
@AskAlexSharov Could you give some links to the code, how to stop the mining then?